flink-kubernetes/src/test/java/org/apache/flink/kubernetes/highavailability/KubernetesHighAvailabilityTestBase.java

```diff
@@ -0,0 +1,127 @@
+
+	protected static final String CLUSTER_ID = "leader-test-cluster";
+	protected static final String LEADER_URL = "akka.tcp://flink@172.20.1.21:6123/user/rpc/resourcemanager";
+	protected static final long TIMEOUT = 30L * 1000L;
+	protected static final String LEADER_CONFIGMAP_NAME = "k8s-ha-app1-resourcemanager";
+	protected final Map<String, KubernetesConfigMap> configMapStore = new HashMap<>();
+	protected final CompletableFuture<FlinkKubeClient.WatchCallbackHandler<KubernetesConfigMap>> configMapsAndDoCallbackFuture =
+		new CompletableFuture<>();
+	protected final CompletableFuture<FlinkKubeClient.WatchCallbackHandler<KubernetesConfigMap>> leaderRetrievalConfigMapCallback =
+		new CompletableFuture<>();
+
+	@Before
+	public void setup() {
+		configuration.setString(KubernetesConfigOptions.CLUSTER_ID, CLUSTER_ID);
+	}
+
+	@After
+	public void teardown() {
+		executorService.shutdownNow();
+	}
+
+	protected KubernetesLeaderElectionService createLeaderElectionService(AtomicBoolean leaderController) {
+		final TestingFlinkKubeClient flinkKubeClient = TestingFlinkKubeClient.builder()
+			.setConfigMapStore(configMapStore)
+			.setWatchConfigMapsAndDoCallbackFunction((ignore, handler) -> {
+				configMapsAndDoCallbackFuture.complete(handler);
+				return new TestingFlinkKubeClient.MockKubernetesWatch();
+			})
+			.setLeaderController(leaderController).build();
+		return new KubernetesLeaderElectionService(
+			flinkKubeClient,
+			executorService,
+			KubernetesLeaderElectionConfiguration.fromConfiguration(LEADER_CONFIGMAP_NAME, configuration));
+	}
+
+	protected KubernetesLeaderRetrievalService createLeaderRetrievalService() {
+		final TestingFlinkKubeClient flinkKubeClient = TestingFlinkKubeClient.builder()
+			.setConfigMapStore(configMapStore)
+			.setWatchConfigMapsAndDoCallbackFunction((ignore, handler) -> {
+				leaderRetrievalConfigMapCallback.complete(handler);
+				return new TestingFlinkKubeClient.MockKubernetesWatch();
+			}).build();
+		return new KubernetesLeaderRetrievalService(flinkKubeClient, LEADER_CONFIGMAP_NAME);
+	}
+
+	/**
+	 * Context to leader election and retrieval tests.
+	 */
+	protected class Context {
+		final AtomicBoolean leaderController = new AtomicBoolean(false);
+		final KubernetesLeaderElectionService leaderElectionService = createLeaderElectionService(leaderController);
+		final TestingContender contender = new TestingContender(LEADER_URL, leaderElectionService);
+
+		final KubernetesLeaderRetrievalService leaderRetrievalService = createLeaderRetrievalService();
+		final TestingListener listener = new TestingListener();
+
+		protected final void runTest(RunnableWithException testMethod) throws Exception {
+			leaderElectionService.start(contender);
+			leaderController.set(true);
+			contender.waitForLeader(TIMEOUT);
+			assertThat(contender.isLeader(), is(true));
+			leaderRetrievalService.start(listener);
+			testMethod.run();
+			leaderElectionService.stop();
+			leaderRetrievalService.stop();
+		}
```

Comment:
I would suggest to separate the action granting leadership to the contender apart from this method, for better readability.

Reply:
Or maybe renaming the method to `runTestAndGrantLeadershipToContender`, if this is needed for almost every test case.